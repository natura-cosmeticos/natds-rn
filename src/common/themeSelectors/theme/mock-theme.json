{
  "avatarSizes": {
    "tiny": {
      "size": 24,
      "fontSize": ".625rem"
    },
    "small": {
      "size": 32,
      "fontSize": ".875rem"
    },
    "standard": {
      "size": 40,
      "fontSize": "1rem"
    },
    "large": {
      "size": 48,
      "fontSize": "1rem"
    },
    "huge": {
      "size": 80,
      "fontSize": "1.5rem"
    }
  },
  "buttonSizes": {
    "small": {
      "paddingTop": 8,
      "paddingRight": 8,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "minHeight": 32
    },
    "medium": {
      "paddingTop": 12,
      "paddingRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "minHeight": 40
    },
    "large": {
      "paddingTop": 16,
      "paddingRight": 16,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "minHeight": 48
    }
  },
  "fontSize": {
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    },
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    }
  },
  "iconSizes": {
    "micro": 16,
    "tiny": 24,
    "small": 32,
    "standard": 48,
    "medium": 64,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "huge": 144,
    "hugex": 192,
    "hugexx": 256
  },
  "opacity": {
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8,
    "00": 0,
    "04": 0.04,
    "08": 0.08
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "none": 0,
    "circle": "50%",
    "small": 2,
    "medium": 4,
    "large": 8
  },
  "sizes": {
    "none": 0,
    "micro": 4,
    "tiny": 8,
    "small": 16,
    "standard": 24,
    "semi": 32,
    "semix": 40,
    "medium": 48,
    "mediumx": 56,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingNone": 0,
    "spacingMicro": 4,
    "spacingTiny": 8,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingSemi": 32,
    "spacingLarge": 48,
    "spacingHuge": 64
  },
  "colorTokens": {
    "colorPrimary": "#F4AB34",
    "colorOnPrimary": "#333333",
    "colorPrimaryLight": "#FEFDE8",
    "colorOnPrimaryLight": "#333333",
    "colorPrimaryDark": "#EF8426",
    "colorOnPrimaryDark": "#333333",
    "colorSecondary": "#FF6B0B",
    "colorOnSecondary": "#000000",
    "colorSecondaryLight": "#FFF8E1",
    "colorOnSecondaryLight": "#000000",
    "colorSecondaryDark": "#FF5808",
    "colorOnSecondaryDark": "#000000",
    "colorBackground": "#FAFAFA",
    "colorOnBackground": "#333333",
    "colorSurface": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorHighlight": "#000000",
    "colorHighEmphasis": "#333333",
    "colorMediumEmphasis": "#777777",
    "colorLowEmphasis": "#BBBBBB",
    "colorLink": "#227BBD",
    "colorOnLink": "#FFFFFF",
    "colorSuccess": "#569A32",
    "colorOnSuccess": "#FFFFFF",
    "colorWarning": "#FCC433",
    "colorOnWarning": "#333333",
    "colorAlert": "#E74627",
    "colorOnAlert": "#FFFFFF"
  },
  "elevation": {
    "1": {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 1, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 1
    },
    "2": {
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 2, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 2
    },
    "3": {
      "elevation": 3,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 3, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 4
    },
    "4": {
      "elevation": 4,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 4, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 5
    },
    "5": {
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 6, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "6": {
      "elevation": 6,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 8, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "7": {
      "elevation": 7,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 9, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 12
    },
    "8": {
      "elevation": 8,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 12, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 17
    },
    "9": {
      "elevation": 9,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 16, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 24
    },
    "10": {
      "elevation": 10,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 24, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 38
    }
  }
}
