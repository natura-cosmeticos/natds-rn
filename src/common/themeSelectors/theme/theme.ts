import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';
import { IElevation } from '@naturacosmeticos/natds-styles/dist/tokens/elevation/IElevation';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutColorTokens = Omit<TokensWithoutColor, 'colorTokens'>
type TokensWithoutElevation = Omit<TokensWithoutColorTokens, 'elevation'>

interface ElevationProps {
  elevation: number
  shadowColor: string
  shadowOffset: { height: number, width: number }
  shadowOpacity: number
  shadowRadius: number
}

type Elevation = Record<keyof IElevation, ElevationProps>


export type Theme = TokensWithoutElevation
  & {
    colorTokens: IColorThemeTokens,
    elevation: Elevation | {}
  }

type ThemeMode = 'dark' | 'light';

const parseElevation = (elevation: string) => {
  const regEx = /(\d+(\.\d+)?)/g;
  const matches = elevation.match(regEx);

  if (matches) {
    const [width, height, blurRadius, , r, g, b, alpha] = matches;

    const toUnit = (prop: string) => parseInt(prop, 0);

    return {
      elevation: 2,
      shadowColor: `rgb(${r}, ${g}, ${b})`,
      shadowOffset: { height: toUnit(height), width: toUnit(width) },
      shadowOpacity: parseFloat(alpha),
      shadowRadius: toUnit(blurRadius),
    };
  }

  return {};
};

const buidElevation = (elevation: IElevation) => {
  const nativeElevation = {};

  Object
    .keys(elevation)
    .forEach((key) => { nativeElevation[key] = parseElevation(elevation[key]); });

  return nativeElevation;
};

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
    elevation: buidElevation(localTokens.elevation),
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
