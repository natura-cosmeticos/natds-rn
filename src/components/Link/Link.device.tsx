import { storiesOf } from '@storybook/react-native';
import { Default, Types, Interactive } from './Link.stories';

storiesOf('Link', module)
  .add('Default', Default)
  .add('Types', Types)
  .add('Interactive', Interactive);
