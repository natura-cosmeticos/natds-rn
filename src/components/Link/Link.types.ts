import { ReactNode } from 'react';
import { NativeSyntheticEvent, NativeTouchEvent } from 'react-native';

export type LinkTypes = 'standard' | 'underline';

export interface LinkProps {
  /**
   * The content of the link.
   * @required
   */
  children: ReactNode;
  /**
   * Handler to be called when the user taps the link.
   * @required
   */
  onPress: (ev: NativeSyntheticEvent<NativeTouchEvent>) => void;
  /**
   * Optional ID for testing.
   */
  testID?: string;
  /**
   * Controls when the link should have an underline.
   *
   * - Available types: `standard`, `underline`.
   * @default 'standard'
   */
  type?: LinkTypes;
}
