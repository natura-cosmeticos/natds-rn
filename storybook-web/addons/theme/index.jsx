import React, { useEffect, Fragment } from 'react';
import { useAddonState, Consumer } from '@storybook/api';
import { TabButton, WithTooltip, TooltipLinkList } from '@storybook/components';
import {
  PANEL_ID, CHANGE, PARAM_KEY, getStoryBookTheme,
} from './shared';
import { light, dark } from '../../theme';

import './styles.css';

const DEFAULT_THEME = {
  name: 'natura',
  type: 'light',
};

const buildThemeProps = mode => ({ ...DEFAULT_THEME, type: mode });

function parseTheme(themes) {
  return Object.entries(themes).reduce((accum, [name, variants]) => {
    Object.entries(variants).forEach(([type]) => {
      accum.push({ name, type });
    });

    return accum;
  }, []);
}

const createBackgroundSelectorItem = ({ name, type }, onChange, selectedItem) => ({
  active: selectedItem.name === name && selectedItem.type === type,
  id: `${name} - ${type}`,
  onClick: () => onChange({ name, type }),
  title: `${name} - ${type}`,
});

const getDisplayedItems = (list, onChange, selectedItem) => {
  if (!list.length) {
    return [];
  }

  return [
    ...list.map(params => createBackgroundSelectorItem(params, onChange, selectedItem)),
  ];
};

const mapper = ({ api, state }) => {
  const story = state.storiesHash[state.storyId];
  const { themes = {} } = (story && api.getParameters(story.id, PARAM_KEY)) || {};

  return { items: parseTheme(themes) };
};

export default function Theme({ channel, api }) {
  const apiTheme = getStoryBookTheme();
  const initialTheme = apiTheme ? buildThemeProps(apiTheme) : DEFAULT_THEME;
  const [currentTheme, changeTheme] = useAddonState(PANEL_ID, initialTheme);
  const { disabled } = api.getCurrentParameter(PARAM_KEY) || {};

  const handleChange = (params) => {
    changeTheme(params);
    api.setOptions({
      theme: params.type === 'light' ? light : dark,
    });
    const event = new Event('CHANGE_THEME', { detail: params.type });

    window.dispatchEvent(event);
    channel.emit(CHANGE, params);
  };

  const renderTooltip = items => ({ onHide }) => (
    <TooltipLinkList
      links={getDisplayedItems(items, (index) => {
        handleChange(index);
        onHide();
      }, currentTheme)}
    />
  );

  const eventListener = event => handleChange(buildThemeProps(event.matches ? 'dark' : 'light'));

  useEffect(() => {
    handleChange(initialTheme);

    const matchMediaEnvent = window.matchMedia('(prefers-color-scheme: dark)');

    matchMediaEnvent.addEventListener('change', eventListener);

    return () => {
      matchMediaEnvent.removeEventListener('change', eventListener);
    };
  }, []);

  if (disabled) return null;

  return (
    <Consumer filter={mapper}>
      {({ items }) => (items.length ? (
          <Fragment>
            <WithTooltip
              placement="top"
              trigger="click"
              closeOnClick
              tooltip={renderTooltip(items)}
            >
              <TabButton
                key="Theme"
                title="Change the Theme">
                {currentTheme.name} - {currentTheme.type}
              </TabButton>
            </WithTooltip>
          </Fragment>
      ) : null)}
    </Consumer>
  );
}
