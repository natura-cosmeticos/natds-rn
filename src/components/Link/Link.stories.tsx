import React from 'react'
import { Alert } from 'react-native'
import { select, text as textKnob } from '@storybook/addon-knobs'

import { StoryContainer, StoryWrapper } from '../../common/HelperComponents/StoryContainer'
import { Link } from './Link'
import { LinkTypes } from './Link.types'

const description = () => `
---

**NOTE**: This component is available in the following variants:

  - ✅ standard

With the following attribute status:

- **Type:**
    - ✅ \`standard\`
    - ✅ \`underline\`

---
`

export default {
  component: Link,
  parameters: {
    componentSubtitle: 'The link component allows you to have a pattern for anchor text.',
    docs: {
      extractComponentDescription: description
    }
  },
  title: 'Components|Link'
}

const onPress = () => {
  Alert.alert('something happened')
}

export const Default = () => (
  <StoryContainer title="Standard">
    <Link onPress={onPress}>Natura Design System</Link>
  </StoryContainer>
)

export const Variants = () => (
  <StoryWrapper title="Variants">
    <StoryContainer title="Standard">
      <Link onPress={onPress}>Natura Design System</Link>
    </StoryContainer>
    <StoryContainer title="Underline">
      <Link onPress={onPress} type="underline">Natura Design System</Link>
    </StoryContainer>
  </StoryWrapper>
)

export const Interactive = () => (
  <StoryContainer title="Interactive">
    <Link
      onPress={onPress}
      type={select('Type', ['standard', 'underline'], 'standard') as LinkTypes}
    >
      {textKnob('Children', 'Natura Design System')}
    </Link>
  </StoryContainer>
)
