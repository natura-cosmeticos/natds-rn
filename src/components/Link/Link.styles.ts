import styled from 'styled-components/native';

import { Theme, getColorLink } from '../../common/themeSelectors';
import { LinkProps } from './Link.types';

type LinkStyleProps = { theme: Theme } & Pick<LinkProps, 'type'>;

export const Label = styled.Text<LinkStyleProps>(({ theme, type }: LinkStyleProps) => ({
  color: getColorLink(theme),
  textDecorationLine: type === 'underline' ? 'underline' : 'none',
}));
