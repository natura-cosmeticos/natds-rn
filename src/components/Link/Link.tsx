import React from 'react';
import { LinkProps } from './Link.types';
import { Label } from './Link.styles';

export const Link = ({
  children,
  onPress,
  testID,
  type = 'standard',
}: LinkProps) => (
  <Label
    testID={testID}
    onPress={onPress}
    type={type}
  >
    {children}
  </Label>
);
