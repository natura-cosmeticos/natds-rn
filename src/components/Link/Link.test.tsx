import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { ThemeProvider } from 'styled-components/native';
import theme from '../../common/themeSelectors/theme/mock-theme.json';
import { Link } from './Link';
import { LinkProps } from './Link.types';

jest.mock(
  'react-native/Libraries/Components/Touchable/TouchableOpacity',
  () => 'TouchableOpacity',
);

const link = (props: Omit<LinkProps, 'theme'>) => (
  <ThemeProvider theme={theme}>
    <Link {...props}>{props.children}</Link>
  </ThemeProvider>
);

const defaultProps = {
  children: 'something',
  onPress: jest.fn(),
  testID: 'link',
};

describe('Link component', () => {
  it('should render the correct color token', () => {
    const { queryByTestId } = render(link(defaultProps));

    expect(queryByTestId('link')).toHaveStyle({ color: '#227bbd' });
  });
  it('should render with underline style', () => {
    const { queryByTestId } = render(
      link({ ...defaultProps, type: 'underline' }),
    );

    expect(queryByTestId('link')).toHaveStyle({
      textDecorationLine: 'underline',
    });
  });
  it('should render without underline style as default', () => {
    const { queryByTestId } = render(link(defaultProps));

    expect(queryByTestId('link')).toHaveStyle({
      textDecorationLine: 'none',
    });
  });
  it('should call the given onPress function', () => {
    const onPress = jest.fn();
    const { getByTestId } = render(link({ ...defaultProps, onPress }));

    fireEvent.press(getByTestId('link'));

    expect(onPress).toHaveBeenCalled();
  });
  it('should take a snapshot', () => {
    const component = render(link(defaultProps)).toJSON();

    expect(component).toMatchSnapshot();
  });
});
