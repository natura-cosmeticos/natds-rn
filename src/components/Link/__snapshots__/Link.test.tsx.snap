// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should take a snapshot 1`] = `
<Text
  onPress={[MockFunction]}
  style={
    Array [
      Object {
        "color": "#227bbd",
        "textDecorationLine": "none",
      },
    ]
  }
  testID="link"
  type="standard"
>
  something
</Text>
`;
